qiskit/qasm3/exporter.py

```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
         )
 
 
-class GlobalNamespace:
-    """Global namespace dict-like."""
-
-    BASIS_GATE = object()
-
-    qiskit_gates = {
-        "p": standard_gates.PhaseGate,
-        "x": standard_gates.XGate,
-        "y": standard_gates.YGate,
-        "z": standard_gates.ZGate,
-        "h": standard_gates.HGate,
-        "s": standard_gates.SGate,
-        "sdg": standard_gates.SdgGate,
-        "t": standard_gates.TGate,
-        "tdg": standard_gates.TdgGate,
-        "sx": standard_gates.SXGate,
-        "rx": standard_gates.RXGate,
-        "ry": standard_gates.RYGate,
-        "rz": standard_gates.RZGate,
-        "cx": standard_gates.CXGate,
-        "cy": standard_gates.CYGate,
-        "cz": standard_gates.CZGate,
-        "cp": standard_gates.CPhaseGate,
-        "crx": standard_gates.CRXGate,
-        "cry": standard_gates.CRYGate,
-        "crz": standard_gates.CRZGate,
-        "ch": standard_gates.CHGate,
-        "swap": standard_gates.SwapGate,
-        "ccx": standard_gates.CCXGate,
-        "cswap": standard_gates.CSwapGate,
-        "cu": standard_gates.CUGate,
-        "CX": standard_gates.CXGate,
-        "phase": standard_gates.PhaseGate,
-        "cphase": standard_gates.CPhaseGate,
-        "id": standard_gates.IGate,
-        "u1": standard_gates.U1Gate,
-        "u2": standard_gates.U2Gate,
-        "u3": standard_gates.U3Gate,
-    }
-    include_paths = [abspath(join(dirname(__file__), "..", "qasm", "libs"))]
+# Just needs to have enough parameters to support the largest standard (non-controlled) gate in our
+# standard library.  We have to use the same `Parameter` instances each time so the equality
+# comparisons will work.
+_FIXED_PARAMETERS = (Parameter("p0"), Parameter("p1"), Parameter("p2"), Parameter("p3"))
+
+# Mapping of symbols defined by `stdgates.inc` to their gate definition source.
+_KNOWN_INCLUDES = {
```

Comment:
The mapping between Qiskit's gate names and the name that they're given in OQ2 and OQ3 isn't entirely identical (Qiskit's OQ2 `qelib.inc` extension calls `c3sx` something different, iirc), and there's the aliasing changes, so I wouldn't necessarily say that there's a duplication in sources of truth here; the "linking" between Qiskit's source of truth and OpenQASM's is itself a new set of facts to be encoded.

Personally I wrote it the way I did because I find the (relatively short) table easier to read and understand than generation code, and there's less to debug in terms of logic.  The more code I write, I find that I'm raising my internal barrier to wanting to write "code that writes code" and so on - I used to do it a lot more, then it was often more effort to maintain than a slightly longer but more explicit list.

Obviously there's a limit to that - I'm of course not suggesting that tablegen has no purpose.

Reply:
Sure, those are good arguments. Simple, readable, self-contained code is nice to find.